compiler/z/codegen/BinaryEvaluator.cpp

```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
+            } 
 
-         int32_t tZeros = trailingZeroes(value);
-         int32_t lZeros = leadingZeroes(value);
-         int32_t tOnes  = trailingZeroes(~value);
-         int32_t lOnes  = leadingZeroes(~value);
-         int32_t popCnt = populationCount(value);
-         int32_t msBit = -1;
-         int32_t lsBit = -1;
+         int32_t tZeros = trailingZeroes(longConstValue);
+         int32_t lZeros = leadingZeroes(longConstValue);
+         int32_t tOnes  = trailingZeroes(~longConstValue);
+         int32_t lOnes  = leadingZeroes(~longConstValue);
+         int32_t popCnt = populationCount(longConstValue);
+
+         // the minimum value for the bit positions is 0, and lsBit can equal msBit iff popCnt = 1
+         // if popCnt == 0 then we cannot generate a RISBG instruction because RISBG preserves at least one bit position
+         int32_t msBit = 0;           
+         int32_t lsBit = 0;  
 
          if (comp->getOption(TR_TraceCG))
             {
             traceMsg(comp,"[%p] land => rotated-and-insert: tZeros %d, lZeros %d, popCnt %d\n", node, tZeros, lZeros, popCnt);
             traceMsg(comp,"          => rotated-and-insert: lOnes %d, tOnes %d\n", lOnes, tOnes);
             }
 
-         bool doIt = false;
+         bool doTransformation = false;
+
+         // We now try to detect if the lconst's bit pattern is such that we can use a single RISBG
+         // instead of multiple AND instructions
+
+         // The first block will catch all cases where the lConstValue is a series of contiguous ones 
+         // surrounded by zeros
+         // ex: 0001111000, 011110, 111111, 001111, 11100 etc...
+         // 
+         // The 'else if' block will consider all cases where we have a series of contiguous zeros 
+         // surrounded by ones
+         // ex: 1110000111, 1000001
+         //
+         // Note: if the lconst is 0x0 we cannot generate a RISBG instruction.
+         // The starting and ending bit positions are inclusive, so specifying (lsBit,msBit)=(0,0) will
+         // preserve the 0th bit in the firstChild, which would be incorrect.
+         // If such a case arises where we can figure out at compile time that the 
+         // constant's value is 0, then we must handle that case in the simplifier.
          if (popCnt == (64 - lZeros - tZeros))
             {
-            msBit = lZeros;
+            msBit = lZeros; 
             lsBit = 63 - tZeros;
-            if (firstChild->getReferenceCount()>1 || skipedConversion != NULL   // Better of with non-destructive RISBG as AND?
-                  || lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0))  // Better off with a single AND?
+            if (firstChild->getReferenceCount() > 1 // if we use an AND and refCount>1, then we will have to backup the value in another register beforehand, as AND is destructive
```

Comment:
Can't we clobber evaluate `firstChild` and perform the transformation anyway? Can you further explain why reference count matters here?

Reply:
So the RISBG instruction is non-destructive and we can keep the value in a register as is. So if the reference count > 1 we can generate something like this: `RISBGN R1, R2, 0,129,0` in order to preserve the data in R2. I don't understand clobberEvaluate completely, but from what I understand so far it would keep a copy of the data in R2 in a separate register. Isn't this unnecessary since we can use `RISBGN` to not clobber the value in the first place?

This is what my comment on the line kind of refers to as well. If I'm missing something here then let me know. Otherwise I think I'll add a better comment in the code to document this.

Reply:
> So the RISBG instruction is non-destructive and we can keep the value in a register as is. So if the reference count > 1 we can generate something like this: RISBGN R1, R2, 0,129,0 in order to preserve the data in R2. I don't understand clobberEvaluate completely, but from what I understand so far it would keep a copy of the data in R2 in a separate register. Isn't this unnecessary since we can use RISBGN to not clobber the value in the first place?

That's correct. Then there is no need to clobber evaluate in this case at all. I don't see why the reference count would then be a pre-requisite for this optimization?

Reply:
If we were to remove the `firstChild->getReferenceCount()>1` condition, the inner if clause will become: `if (lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0)`. 

We can still have a case where the outer if clause (  `if (popCnt == (64 - lZeros - tZeros))`) is true, but the new inner one is not (ex. if `lConstValue` is originally `0x0000FFFFFFFFFFFF`). In this scenario we would bail from the helper and end up generating an `NIHH` instruction. If the firstChild's refCount>1, we will need to clobber evaluate this case because we cannot destroy the value. Wouldn't this result in an extra instruction since we need to backup the original value somewhere? 
If we went with a `RISBG` instruction instead, we can accomplish this in one instruction since we don't need to clobber evaluate. The number of registers used would be the same in both cases. This is why I believe if the refCount>1 for the firstChild we should always use RISBG since we save an instruction.

I'm going to test this case out once I get past the compile failures I'm currently seeing. Let me know if I'm missing something here.

Reply:
> We can still have a case where the outer if clause ( if (popCnt == (64 - lZeros - tZeros))) is true, but the new inner one is not (ex. if lConstValue is originally 0x0000FFFFFFFFFFFF). In this scenario we would bail from the helper and end up generating an NIHH instruction. If the firstChild's refCount>1, we will need to clobber evaluate this case because we cannot destroy the value. Wouldn't this result in an extra instruction since we need to backup the original value somewhere?

Right, but we miss all the opportunities where `lConstValue` is not one where we can `NIHH` for example. The point I'm trying to drive is that `firstChild->getReferenceCount() == 1` does not always imply that using `RISBG` is a bad idea. 

What if `firstChild->getReferenceCount() == 1` and `lConstValue == 0x0000FFFFFFFFFFF0` for example? This function would choose not to generate an `RISBG` and the AND evaluator will have to generate at least 2 ANDs.

Reply:
> We can still have a case where the outer if clause ( if (popCnt == (64 - lZeros - tZeros))) is true, but the new inner one is not (ex. if lConstValue is originally 0x0000FFFFFFFFFFFF). In this scenario we would bail from the helper and end up generating an NIHH instruction.

The difference between generating `NIHH` and `RISBG` in this case is that `RISBG` is 2 bytes longer. The alternative to not generating `RISBG` in case of `firstChild->getReferenceCount() == 1` and `lConstValue == 0x0000FFFFFFFFFFF0` is that multiple instructions will get generated, which is always worse than the extra 2 bytes of icache we pay.

The best of both worlds is to determine in which case can we *NOT* use one of the `NI**` variants to evaluate the sequence when `firstChild->getReferenceCount() == 1` and generate an `RISBG` in such cases (if possible of course).

Reply:
Okay so I think we're both trying to say the same thing here.   

>The point I'm trying to drive is that firstChild->getReferenceCount() == 1 does not always imply that using RISBG is a bad idea.

 The `firstChild->getReferenceCount() > 1` check is not a prerequisite. The entire if clause looks like the following in a single line: 
`if (firstChild->getReferenceCount() > 1 || lZeros > 31 || tZeros > 31 || (lZeros > 0 && tZeros > 0))`

I separated these into multiple lines to add comments explaining what each one is doing. I'll clean this up and put a comment above the `if` to make this more clear.

>What if firstChild->getReferenceCount() == 1 and lConstValue == 0x0000FFFFFFFFFFF0

We do handle this case. We would catch it with this `OR` : `(lZeros > 0 && tZeros > 0)`. The reference count is irrelevant here. We will generate RISBG for this case no matter what. 

So if our lConstValue satisfies the pattern that the outer if-clause is trying to detect AND if the refCount>1 then it's ALWAYS better to generate a RISBG instruction. The alternative to a RISBG is multiple instructions (back up the value, then use destructive AND instruction).

>The best of both worlds is to determine in which case can we NOT use one of the NI** variants to evaluate the sequence when firstChild->getReferenceCount() == 1 and generate an RISBG in such cases (if possible of course).

We do essentially do that. The inner-if will only allow a RISBG to be generated if the alternative is to generate multiple `ANDs`. The refCount>1 check is done for the reasons explained above. The remaining checks in this clause can be explained by the following explanation:
>I think a good way to think about this is that we are better off generating a RISBG if there's a 0 in the top half AND the bottom half of the constant (given that we have the leadingOnes or leadingZeros pattern). A 0 means that we need to explicitly zero out a particular bit in the register value if we use these AND Immediate instructions. If the zero exists in both halves then we will need two AND Immediate instructions.

Reply:
Thanks for the explanation. My mistake was that I was reading the code as:

```
if (firstChild->getReferenceCount())
   {
   ...
   }
```